### PR TITLE
Fix NaN gradient at zero for tf.experimental.numpy.sinc

### LINKED ISSUE #127226: tf.keras.ops.numpy.sinc returns NaN gradient at zero where derivative should be zero
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0, also reproduced on tf 2.22.0-dev20260904

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

Python 3.13.5

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

`tf.keras.ops.numpy.sinc` returns `nan` for the first derivative at a point where the normalized sinc function has a smooth extension and the derivative should be finite.

In the reproducer below, the function evaluates `tf.keras.ops.numpy.sinc(2 * t + 1)` at `t = -0.5`, so the argument to `sinc` is exactly zero. The forward value is correctly returned as `1.0`. However, TensorFlow autodiff returns `nan` for the derivative with respect to `t`.

The expected derivative is `0.0`, including the chain rule through `2 * t + 1`.

Expected behavior?

The first derivative of `tf.keras.ops.numpy.sinc(2 * t + 1)` at `t = -0.5` should be `0.0`, not `nan`.

### Standalone code to reproduce the issue

```shell
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["TF_NUM_INTRAOP_THREADS"] = "1"
os.environ["TF_NUM_INTEROP_THREADS"] = "1"

import tensorflow as tf

def target(t):
    return tf.keras.ops.numpy.sinc(2 * t + 1)

x = tf.constant(-0.5, dtype=tf.float64)

with tf.GradientTape() as tape:
    tape.watch(x)
    y = target(x)

actual = tape.gradient(y, x)
expected = 0.0

print("forward:", y.numpy())
print("actual:", actual.numpy())
print("expected:", expected)

if tf.math.is_nan(actual):
    print("BUG REPRODUCED: tf.keras.ops.numpy.sinc returns NaN gradient at zero")
else:
    print("not reproduced")
```

### Relevant log output

```shell
forward: 1.0
actual: nan
expected: 0.0
BUG REPRODUCED: tf.keras.ops.numpy.sinc returns NaN gradient at zero
```

### PR DESCRIPTION
**Pull request by Venkat6871**
Fixes #127226

`tf.experimental.numpy.sinc` returns a NaN gradient when the input is exactly zero.

This change avoids the `0 / 0` path during gradient computation and correctly returns a zero gradient at zero.

Added regression tests for the zero-input gradient and numerical gradient checks.

### CODE REVIEW

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thanks for contributing this fix for `tf.experimental.numpy.sinc`! The tests and approach for avoiding `0 / 0` division are correct.

However, there is a minor optimization we should make to avoid redundant tensor allocations.
The current implementation conditionally checks `x == 0` and allocates `array_ops.ones_like` twice. We can extract the condition and reuse the ones tensor to save memory allocations.

Could you update the `sinc` function to cache the `x == 0` check and the ones tensor like this?

```python
def sinc(x):
  def f(x):
    pi_x = x * np.pi
    is_zero = x == 0
    ones = array_ops.ones_like(x)
    safe_pi_x = array_ops.where_v2(is_zero, ones, pi_x)
    return array_ops.where_v2(
        is_zero, ones, math_ops.sin(safe_pi_x) / safe_pi_x
    )

  return _scalar(f, x, True)
```

Also, it looks like some external CI jobs are failing, but they appear to be unrelated infrastructure timeouts/flakes. Looking forward to reviewing the updated patch!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi `@Venkat6871`, thank you for the PR and for adding regression tests! The mathematical fix to avoid NaNs at `x=0` inside `np_math_ops.sinc` is perfectly correct.

However, the internal automated test suite and the GitHub Linux CI both failed to execute the new test because of missing build dependencies.

In `tensorflow/python/ops/numpy_ops/np_math_ops_test.py`, you added:
```python
from tensorflow.python.eager import backprop
from tensorflow.python.ops import gradient_checker_v2
```

Because TensorFlow relies on strict Bazel `BUILD` requirements, you must also add these modules to the `test` dependencies in the corresponding `BUILD` file.

Please add `"//tensorflow/python/eager:backprop"` and `"//tensorflow/python/ops:gradient_checker_v2"` to the `deps` list of the `tf_py_strict_test(name = "np_math_ops_test", ...)` target in `tensorflow/python/ops/numpy_ops/BUILD`.

Once you push this fix, the CI should turn green and we can proceed with merging. Thanks again!

**Comment by Venkat6871** [comment]
Hi @dmiltr3 ,
Thanks for the review. I have added the requested Bazel dependencies for `backprop` and `gradient_checker_v2` and pushed the changes to the PR branch.

All GitHub CI checks are now passing, including the Linux CPU/CUDA/CUDA13, Windows, and ARM64 jobs.

**Review by dmiltr3 — APPROVED** [review]
Thanks for the contribution! We've verified this locally and all internal presubmit tests passed cleanly. LGTM.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @Venkat6871,

Thank you for your contribution and for implementing the fix for `tf.experimental.numpy.sinc`! The implementation and test cases have been reviewed and approved.

However, recent changes were merged to the upstream `master` branch in `tensorflow/python/ops/numpy_ops/` (updating `np_math_ops.py` and `np_math_ops_test.py`) after this PR branch was branched. Because this branch has diverged from `master`, the automated merge and presubmit testing pipeline encountered conflicts with out-of-date files when attempting to integrate the change into the upstream repository.

Could you please rebase your branch on the latest `master` (or merge `master` into your branch) and push the updated branch?

```bash
git fetch upstream
git checkout fix/127226-sinc-gradient-zero
git rebase upstream/master
git push --force-with-lease origin fix/127226-sinc-gradient-zero
```

Once the branch is synced with the latest upstream `master`, the automated synchronization and presubmit test suite will automatically run against the clean diff and proceed with merging.

Thanks again!

**Comment by djhdusjbd-eng** [comment]
I value this opportunity. Could you provide more details?

**Comment by Venkat6871** [comment]
@dmiltr3 ,
I have completed the tasks you requested.
Thank you!

**Review by dmiltr3 — APPROVED** [review]
Thank you for the contribution and the robust fix with new tests! Tests have passed successfully internally. Approving.
